perf/stability/alertmanager/webhook.go

```diff
@@ -15,26 +15,141 @@
+	alertingStatus = "ALERTING"
+)
+
+// SingleMonitorStatus represents the status of one single monitor
+type SingleMonitorStatus struct {
+	Name        string
+	Status      string
+	Labels      map[string]string
+	Annotations string
+}
+
+func initPromClient(host string) {
+	var err error
+	promclient, err = api.NewClient(api.Config{
+		Address: host,
+	})
+	if err != nil {
+		log.Fatalf("Error creating client: %v\n", err)
+	}
+	v1api = v1.NewAPI(promclient)
+}
+
+func initSpanner() *spanner.Client {
+	projectID = os.Getenv("PROJECT_ID")
+	instance = os.Getenv("INSTANCE")
+	dbName = os.Getenv("DBNAME")
+	mstableName = os.Getenv("MS_TABLE_NAME")
+	clusterName = os.Getenv("CLUSTER_NAME")
+	branch = os.Getenv("BRANCH")
+	testID = os.Getenv("TESTID")
+
+	ctx := context.Background()
+	var err error
+	db := fmt.Sprintf("projects/%s/instances/%s/databases/%s", projectID, instance, dbName)
+	log.Printf("initializing spanner db: %s\n", db)
+	client, err = spanner.NewClient(ctx, db)
+	if err != nil {
+		log.Fatalf("failed to create spanner client: %v", err)
+	}
+	return client
+}
+
+// initMonitorStatus writes initial MonitorStatus to spanner db.
+func initMonitorStatus() {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	log.Println("checking prometheus rules")
+	rules, err := v1api.Rules(ctx)
+	if err != nil {
+		log.Fatalf("error querying Prometheus for alerts: %v\n", err)
+	}
+	var monitorList []SingleMonitorStatus
+	for _, gs := range rules.Groups {
+		for _, rs := range gs.Rules {
+			switch v := rs.(type) {
+			case v1.RecordingRule:
+				continue
+			case v1.AlertingRule:
+				fmt.Printf("adding alerting rule: %s\n", v.Name)
+				status := healthyStatus
+				if len(v.Alerts) != 0 {
+					status = alertingStatus
+				}
+				monitorList = append(monitorList, SingleMonitorStatus{
+					Annotations: v.Annotations.String(),
+					Name:        v.Name,
+					Status:      status,
+				})
+			default:
+				fmt.Printf("unknown rule type %s", v)
+			}
+		}
+	}
+	log.Println("writing initial monitor status to Spanner")
+	if err := writeMonitorStatusToDB(monitorList, false); err != nil {
+		log.Fatalf("failed to initialize monitor status in Spanner: %v", err)
+	}
+}
+
 func webhook(w http.ResponseWriter, r *http.Request) {
 	defer r.Body.Close()
-	fmt.Println("handling")
+	log.Println("Handling new alert webhook")
 	data := template.Data{}
 	if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
 		log.Fatalf(err.Error())
-		return
 	}
-	log.Printf("Alerts: GroupLabels=%v, CommonLabels=%v", data.GroupLabels, data.CommonLabels)
-	//TODO(richardwxn): process alert
+	log.Printf("alerts: GroupLabels=%v, CommonLabels=%v", data.GroupLabels, data.CommonLabels)
+	var monitorList []SingleMonitorStatus
+	var errs error
 	for _, alert := range data.Alerts {
-		log.Printf("Alert: status=%s,Labels=%v,Annotations=%v", alert.Status, alert.Labels, alert.Annotations)
+		ms, err := convertPromAlertToInternalMonitorStatus(alert)
```

Comment:
consider adding link and other details to the spanner table.

Reply:
In the UI, ensure that the UI is color coded.

If anything needs attention it should be RED :-)

Reply:
yes, added different colors in the dashboard, check it out here: http://eng.istio.io/releasequal